planetiler-core/src/main/java/com/onthegomap/planetiler/archive/TileCopy.java

```diff
@@ -0,0 +1,236 @@
+package com.onthegomap.planetiler.archive;
+
+import com.onthegomap.planetiler.config.Arguments;
+import com.onthegomap.planetiler.config.CommonConfigs;
+import com.onthegomap.planetiler.stats.Counter;
+import com.onthegomap.planetiler.stats.ProcessInfo;
+import com.onthegomap.planetiler.stats.ProgressLoggers;
+import com.onthegomap.planetiler.stats.Stats;
+import com.onthegomap.planetiler.util.Gzip;
+import com.onthegomap.planetiler.util.Hashing;
+import com.onthegomap.planetiler.worker.WorkerPipeline;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility to copy/convert tiles and metadata from one archive into another.
+ * <p>
+ * Example usages:
+ *
+ * <pre>
+ * --input=tiles.mbtiles --output=tiles.mbtiles
+ * --input=tiles.mbtiles --output=tiles.pmtiles --skip_empty=false
+ * --input=tiles.pmtiles --output=tiles.mbtiles
+ * --input=tiles.mbtiles --output=tiles/
+ * --input=tiles.mbtiles --output=tiles.json --out_tile_compression=gzip
+ * --input=tiles.mbtiles --output=tiles.csv --out_tile_compression=none
+ * --input=tiles.mbtiles --output=tiles.proto
+ * </pre>
+ */
+public class TileCopy {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TileCopy.class);
+
+  private final TileCopyConfig config;
+
+  private final Counter.MultiThreadCounter tilesWrittenOverall = Counter.newMultiThreadCounter();
+
+
+  TileCopy(TileCopyConfig config) {
+    this.config = config;
+  }
+
+  public void run() throws IOException {
+
+    if (!config.inArchive().exists()) {
+      throw new IllegalArgumentException("the input archive does not exist");
+    }
+
+    config.outArchive().setup(config.force(), config.append(), config.tileWriterThreads());
+
+    final var loggers = ProgressLoggers.create()
+      .addRateCounter("tiles", tilesWrittenOverall::get);
+
+    try (
+      var reader = TileArchives.newReader(config.inArchive(), config.inArguments());
+      var writer = TileArchives.newWriter(config.outArchive(), config.outArguments())
+    ) {
+
+      final TileArchiveMetadata inMetadata = getInMetadata(reader);
+      final TileArchiveMetadata outMetadata = getOutMetadata(inMetadata);
+
+      writer.initialize();
+      try (
+        var rawTiles = reader.getAllTiles();
+        var it = config.skipEmpty() ? rawTiles.filter(t -> t.bytes() != null && t.bytes().length > 0) : rawTiles
```

Comment:
What if the "native order" of the archive is different between the archives? If they are the same I think the approach of iterating and pushing them through output makes sense, but if they are different we can choose the source or target tile order.  So far, all planetiler written in order to the target archive.

Reply:
Also, there's a lot of things we could add to this, we don't need to add all of them in this PR but one interesting one might be setting a bounding box/zoom range. In that case we'd likely want a different enumeration mechanism that lists all the tiles in the order expected by the dest archive, and queries the source archive for it.  Here's an initial version I was playing with a few months ago that iterates in dest archive order, does work in parallel and joins the ordered tiles at the end: 

https://github.com/onthegomap/planetiler/blob/38ca9b4edbdc9bf194ae028233e873ad2e122dbf/planetiler-core/src/main/java/com/onthegomap/planetiler/util/TileArchiveCopier.java#L122

I guess though we'll never be able to query tiles from a json/protobuf/csv input?

Reply:
> What if the "native order" of the archive is different between the archives? ...
Okay, I see what I can do here.

> I guess though we'll never be able to query tiles from a json/protobuf/csv input?
Well yes, it's implemented here as well. Loading all tiles (coord+data) in order is simple and fast - just read the file from top to bottom. Reading tile data by a coord is also implemented but highly inefficient - so that is not super useful but it's implemented for debugging and simple/small tile sets.

> ...enumeration...
I see. That code is helpful. Think I'll tackle it a bit different - since for json/protobuf/csv loading all corrds and then load data by coord is not efficient. Thinking more in the direction of an iterator - also in order to avoid loading all tileCoords/-IDs into memory. For "in-order"-iteration we can - at least for json/protobuf/csv - load the coord and the tile data eagerly. For anything else we would load the tile data lazy.

Reply:
Got it, seems like we might want to have 2 iteration modes:

- if input/output preferred modes are different AND input format supports random access (and user opts-in?) then enumerate the tiles in preferred output order and query input archive for them
- otherwise use this simple iterator approach

In any case, seems like it might be a good idea to log a warning when the output order is not going to be the preferred order of the output format just so the user is aware. Some formats care more than others, for example pmtiles sets a "clustered=true" header field when output tiles are ordered that enables some client-side query optimizations.

I also think it's fine to start with the simple iterator approach then add the random-access-output-order approach in a later PR as an optimization.

Reply:
Or another option besides random access queries would be to use `ExternalMergeSort` class to write tiles to temporary storage blocks, sort them in parallel, then iterate through them in order to write to the output format. That would work for any input format.

Another complication will be an HTTP tile archive that only supports random access but not iteration (s3 would be similar but does support iteration), but we could cross that bridge when we get there.

I think this just builds the case for doing the output-order change in a separate PR since there are a few options to eva

Reply:
Good timing! @bdon just did a blog post on the benefits of clustered pmtiles archives too: https://protomaps.com/blog/pmtiles-cluster/